### ISSUE TITLE
std::filesystem support

### ISSUE THREAD

**Issue by chenxiaolong** [android/ndk#609]
#### Description

The libc++ libraries (both shared and static) seem to be missing symbols needed to use `<experimental/filesystem>`. For example, trying to compile a simple program that calls `std::experimental::filesystem::remove_all(path)` will fail at link time with:

```
/stuff/Applications/Linux/android-ndk/sources/cxx-stl/llvm-libc++/include/experimental/filesystem:1695: error: undefined reference to 'std::experimental::filesystem::v1::__remove_all(std::experimental::filesystem::v1::path const&, std::__ndk1::error_code*)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

Sample project: https://github.com/chenxiaolong/ndk_filesystem_test (run `./test.sh` to reproduce issue)

Is this expected to happen? I couldn't find any details as to whether the filesystem library should work with the NDK.

#### Environment Details

* NDK Version: r16b (`16.1.4479499`)
* Build sytem: any
* Host OS: Fedora 27 x86_64
* Compiler: clang
* ABI: any
* STL: libc++
* NDK API level: 21
* Device API level: n/a

**Comment by jmgao** [comment]
libc++ compiles its experimental parts into libc++experimental, a library that you need to [link in separately](https://libcxx.llvm.org/docs/UsingLibcxx.html#using-libc-experimental-and-experimental). Unfortunately, we don't seem to build it as part of the NDK. Retitling the bug to look into shipping it.

In the mean time, you might be able to just add the required files directly into your project (from `$ANDROID_NDK_ROOT/sources/cxx-stl/llvm-libc++/src/experimental/filesystem`) to work around this on your end, but I haven't tried it (and this is guaranteed to break in the future, when the files are moved into the standard or abandoned).

**Comment by chenxiaolong** [comment]
Ah okay, that makes perfect sense. I didn't realize libc++ also had a separate library like gcc's libstdc++fs.

**Comment by DanAlbert** [comment]
The tests for `std::filesystem` are not going to be easy to run on Android (they require a python helper). Going to find a way to work around that, but it's not likely to happen during r17.

**Comment by DanAlbert** [comment]
Unlikely to fit into r18. I want to get this done for the platform before the NDK since it should be a smaller problem to solve, and haven't done that yet.

**Comment by andreya108** [comment]
So, for now, can we suppose that full support for std::filesystem will appear in r19?

**Comment by DanAlbert** [comment]
Given how strongly people feel about https://github.com/android-ndk/ndk/issues/463 I will be doing that before this, and I'm skeptical that I'll be able to accomplish that before r19 is complete (there's always a fair amount of overhead in coordinating with external products, and CMake is an unfamiliar code base to all of us here).

Going to move this to the r20 bucket to match the other bug, but odds are there will be some down time while waiting on reviews for #463 so I'll try to make some progress on this then.

**Comment by DanAlbert** [comment]
> So, for now, can we suppose that full support for std::filesystem will appear in r19?

Also, a disclaimer that when this *does* land, the NDK's support for `std::filesystem` will be as complete as libc++'s. The [libc++ docs](https://libcxx.llvm.org/docs/UsingLibcxx.html#using-filesystem-and-libc-fs) still indicate that this is experimental (and not ABI stable), but it does at least appear to be *complete*.

**Comment by arcanis** [comment]
Of note: `<filesystem>` is now available without going through `experimental` (cf [commit](https://llvm.org/viewvc/llvm-project/libcxx/trunk/include/filesystem?view=log)). Would that make it easier to integrate?

**Comment by feliwir** [comment]
I'd like to see the filesystem support happening aswell :+1:

**Comment by diegurrio** [comment]
Looking forward to the experimental file system libraries being available.

**Comment by KayEss** [comment]
In the current NDK the experimental filesystem is still not part of the libc++ library, and now the sources can no longer be installed through the SDK tools so our old work around suggested by @jmgao no longer works.

Any suggestions for what to do now?

**Comment by alexcohn** [comment]
@KayEss you can probably follow https://android.googlesource.com/platform/external/libcxx/+/master/CMakeLists.txt

**Comment by enh** [comment]
https://issuetracker.google.com/121291977 was another request for std::filesystem...

**Comment by arcanis** [comment]
@DanAlbert Is it still planned for r20?

**Comment by DanAlbert** [comment]
This is made slightly more clear on our roadmap: https://android.googlesource.com/platform/ndk/+/master/docs/Roadmap.md#future-work

Essentially, this far out the milestone is just what I'm hoping to do. As the roadmap shows, this is second on my to-do list. Since I haven't gotten far enough on the first item to know how much time it will take, the most accurate thing I can say is that I'll start working on this when I'm done with that. I wouldn't make plans around it being available to apps in May.

As always, I'll continue moving bugs between milestones when it becomes more clear what will and what won't make it.

**Comment by Zingam** [comment]
@DanAlbert The roadmap doesn't make it clear that the items are ordered according to importance. The first item on that list is "cdep"? Are you sure you want to integrate "cdep" first? What about "vcpkg" instead. It supports linux,windows,osx and has already a bunch of libraries.

**Comment by DanAlbert** [comment]
I'll look in to getting the design doc published. `vcpkg` was something I looked in to but it didn't fit all (or even most) of our criteria. It should be worth noting that while it has quite a large corpus, none of those libraries bare for Android.

**Comment by DanAlbert** [comment]
@Zingam https://github.com/android-ndk/ndk/issues/916 explains the current plan.

**Comment by rcdailey** [comment]
I want to make sure I understand if this issue is related to what I'm seeing. I'm using NDK r19c, and I enabled C++17. I try to use `std::filesystem`, but I get linker errors:

```
C:/android/android-ndk-r19c/toolchains/llvm/prebuilt/windows-x86_64/sysroot/usr/include/c++/v1/filesystem:1113: error: undefined reference to 'std::__ndk1::__fs::filesystem::path::__root_directory() const'
  C:/android/android-ndk-r19c/toolchains/llvm/prebuilt/windows-x86_64/sysroot/usr/include/c++/v1/filesystem:1125: error: undefined reference to 'std::__ndk1::__fs::filesystem::path::__filename() const'
```

Is this due to no official filesystem support in NDK r19c? I am not sure I understand the issue here. If I use r20 will filesystem work?

**Comment by DanAlbert** [comment]
`std::filesystem` is not available in the NDK yet. It won't be for some time because there are higher priority tasks in the way.

**Comment by rcdailey** [comment]
Feels like the NDK is struggling to keep up with iterations of C++. I was on r17 for a while trying to use C++17 but a lot of things weren't available (I think `std::result_of` comes to mind, but i don't remember for sure). So I upgraded to r19 for C++17 but seems to be *still* incomplete. Between Windows, Ubuntu, and Android, the NDK is the bottleneck in terms of full C++17 support. I use the same version of Clang on Ubuntu and it seems to be more feature complete. Rather, maybe the disparity is between LLVM STL and not necessarily the compiler. What version of LLVM STL is available in r19? I am not sure how to find that out.

Anyway, for this issue, I'll replace my `std::filesystem` usage with `boost::filesystem` for now. Thanks for letting me know. Is there a way to keep track of C++ standard support progress in the NDK other than following individual issues like this?

**Comment by DanAlbert** [comment]
(unplanned doesn't actually mean unplanned, it's just not planned for a specific release atm because I don't know how long the other things in front of it will take)

> Feels like the NDK is struggling to keep up with iterations of C++.

I hear you. We're good about tracking the trivial updates that just require a `git merge` from upstream, but unfortunately `std::filesystem` requires us to spend time on making those tests runnable in a cross-compiled build. We also have to get them to pass if they don't already, but right now we can't even run them because they were not built to run in a cross-compiled environment, and even less so one where the target doesn't have Python installed.

> Rather, maybe the disparity is between LLVM STL and not necessarily the compiler.

Both happen. `std::filesystem` is waiting on us, but it is more often the case that upstream just doesn't have the functionality yet (it's a fairly small number of people that work on any of the STLs, so they're more often than not overloaded).

libc++ has https://libcxx.llvm.org/docs/#c-dialect-support and Clang has http://clang.llvm.org/cxx_status.html for tracking standard implementation status.

> What version of LLVM STL is available in r19? I am not sure how to find that out.

I was about to point at the changelog, but it looks like I forgot to put that in the r19 changelog :(

Can look up the tag and follow the log from there: https://android.googlesource.com/platform/external/libcxx/+/refs/tags/ndk-r19c leads to https://android.googlesource.com/platform/external/libcxx/+/ffc08e52a5b16f03d8d4923ee544f69b75e47a4d, so r345762.

**Comment by KayEss** [comment]
What I've been doing is to grab the filesystem source for the same version of the compiler and putting the source code into a sub-project to use. It's here: https://github.com/wfp/MDCA-Platform/tree/master/mdca-android/src/filesystem

We were using Boost filesystem as an alternative, but experienced bugs with directory iteration that went away by just using the clang version.

It is a bit of a pain trying to find the right source code version when the NDK is updated. The sources used to ship with the NDK and then it was quite simple.

Is there any way to just have an unsupported build or sources for these files included somewhere to make this process easier?

**Comment by chussong** [comment]
> What I've been doing is to grab the filesystem source for the same version of the compiler and putting the source code into a sub-project to use. It's here: https://github.com/wfp/MDCA-Platform/tree/master/mdca-android/src/filesystem

We've done the same thing and also found it a workable-but-annoying workaround.

I also noticed that [LLVM 9.0 merges libc++fs into main libc++](https://libcxx.llvm.org/docs/UsingLibcxx.html#using-filesystem). It's scheduled for a final release next month; I wondered if this had any bearings on your plans to ship <filesystem> support for Android.

**Comment by DanAlbert** [comment]
As I've said several times in this thread: the problem is testing and has nothing to do with upstream. We need to invest time in getting this tested and we don't currently have the time available.

**Comment by gitoby** [comment]
I am one of the authors of the Boden cross-platform framework (www.boden.io). We're trying to make it easier for C++ developers to make mobile apps and we're also hitting this issue currently.

Could you provide us with a rough estimate on when this may be done? Something along the lines of "won't happen this year" would already be helpful to us.

Also, if there's anything we can do to speed up the process on this issue, please let me know.

**Comment by DanAlbert** [comment]
Fortunately some of the current work is finishing up, so someone should be able to at least look at this soon. Note that https://github.com/android-ndk/ndk/issues/463 will have to be delayed for this.

> Could you provide us with a rough estimate on when this may be done? Something along the lines of "won't happen this year" would already be helpful to us.

In *a* release this year should be doable, barring any unforeseen difficulties. Whether or not it's available in a stable release depends a lot on timing. I would really like it to be in r21, but I think we'd be delaying r21 about a quarter to make that happen... We should know more precisely when someone has a chance to start digging in (probably a month from now, maybe sooner).

> Also, if there's anything we can do to speed up the process on this issue, please let me know.

Not sure what you had in mind, but the NDK is open source. If that's something you're interested in, I can link the docs for building/testing the NDK and go over what I *think* fixing this bug will entail.

**Comment by gitoby** [comment]
Thanks a lot for your detailed response. This really helps us with our release planning and technical decision making.

> Not sure what you had in mind, but the NDK is open source. If that's something you're interested in, I can link the docs for building/testing the NDK and go over what I think fixing this bug will entail.

Yes, that's kind of what I had in mind. I cannot promise anything at this point, but if you could provide us with the necessary information we can have a look and discuss whether we'd be able to allocate the resources to work on this.

Again, thanks for your help!

**Comment by DanAlbert** [comment]
Okay, great! Here's a handful of docs to get you started:

If you don't have `repo` already, follow https://source.android.com/setup/build/downloading to get set up.

https://android.googlesource.com/platform/ndk/+/master/docs/Building.md covers building the NDK. There's a docker file at https://android.googlesource.com/platform/ndk/+/master/infra/docker/Dockerfile that can help if you're having trouble getting all the right dependencies installed.

https://android.googlesource.com/platform/ndk/+/master/docs/Testing.md describes testing. Note that `run_tests.py` needs to be run with python 3.6 or newer. Can run just the libc++ tests with `./run_tests.py --rebuild --filter libc++` (can filter further as needed).

https://source.android.com/setup/contribute/submit-patches.html explains how to upload your patch (at some point in that process, gerrit will have you sign the Google CLA if you haven't already).

The libc++ source is in external/libcxx. I'll get that updated at some point today so you've at least got the latest source to work with.

I think what roughly needs to be done is:

- [ ] Add the `std::filesystem` sources to the libc++ module in https://android.googlesource.com/platform/external/libcxx/+/refs/heads/ndk/Android.mk.
- [ ] Configure the test runner to actually run the tests. See `configure_features` in https://android.googlesource.com/platform/external/libcxx/+/refs/heads/ndk/utils/libcxx/ndk/test/config.py#86.
- [ ] Fix any configuration of the libc++ test runner in https://android.googlesource.com/platform/ndk/+/refs/heads/master/ndk/test/types.py#571. The `std::filesystem` tests require a handful of extra inputs for some input test data, as well as an external program that is used for generating test data dynamically from the tests.
- [ ] Get the test runner to sync the test data to the device. There's already some support for this, but I think it needs some tweaks for the `std::filesystem` tests because the layout is different.
- [ ] Get that external program working. https://android.googlesource.com/platform/external/libcxx/+/refs/heads/ndk/test/support/filesystem_dynamic_test_helper.py and https://android.googlesource.com/platform/external/libcxx/+/refs/heads/ndk/test/support/filesystem_test_helper.hpp are the current implementation.
- [ ] Fix any broken tests. This wasn't too hard when I did it for the OS, but I don't know what to expect for Jelly Bean.

The file system test tool is the part that's going to be a pain. By default this is Python, but getting Python running on Jelly Bean is probably more work than should be needed here. Using a shell script might be easiest, but we need the tests to run on Jelly Bean which was missing a *lot* of what you'd expect from a shell environment. Might be able to build toybox for Jelly Bean to cover the gaps... Alternatively this could be rewritten to not use an external tool. I believe upstream's reasoning for using Python was that it's safe to assume that any system they're testing on has a Python with functioning file system APIs, and they wanted to isolate the test tools from `std::filesystem` but also keep it relatively portable. Adding a fall-back mode that just uses the POSIX filesystem APIs rather than an external tool would probably be a reasonable approach. I think @enh-google had actually taken a look at this yesterday, but I'm not sure how far he got.

We also need to decide what to do about `_FILE_OFFSET_BITS`. I'm pretty sure we just need to ignore this and not support large file offsets for LP32 initially, since there's only one libc++ per ABI, but it would be worth checking which parts of the implementation actually use `off_t` and see when the needed APIs were added to bionic. If there's no easy answer we can punt to a later release.

LMK if you have any questions I didn't cover, and if the above sounds like something you're interested in tackling.

**Comment by enh-google** [comment]
> I think @enh-google had actually taken a look at this yesterday, but I'm not sure how far he got.

```
  Expected Passes    : 181
  Expected Failures  : 1
  Unsupported Tests  : 4
  Unexpected Failures: 2
```

one of those is trying to use too much disk space though i can't actually see what the failure is; i just get
```
******************** TEST 'libc++ :: std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_large.pass.cpp' FAILED ********************
Compiled With: 'prebuilts/clang/host/linux-x86/clang-r353983d/bin/clang++ -c -Wno-null-pointer-arithmetic -Wno-null-dereference -Iexternal/libcxx/test/support -Iexternal/libcxx -mthumb -Os -fomit-frame-pointer -DANDROID -fmessage-length=0 -W -Wall -Wno-unused -Winit-self -Wpointer-arith -no-canonical-prefixes -DNDEBUG -UDEBUG -fno-exceptions -Wno-multichar -O2 -g -fno-strict-aliasing -fdebug-prefix-map=/proc/self/cwd= -D__compiler_offsetof=__builtin_offsetof -faddrsig -Wimplicit-fallthrough -Wno-error=implicit-fallthrough -Werror=int-conversion -fexperimental-new-pass-manager -Wno-reserved-id-macro -Wno-format-pedantic -Wno-unused-command-line-argument -fcolor-diagnostics -Wno-zero-as-null-pointer-constant -Wno-sign-compare -Wno-defaulted-function-deleted -Wno-inconsistent-missing-override -ffunction-sections -fdata-sections -fno-short-enums -funwind-tables -fstack-protector-strong -Wa,--noexecstack -D_FORTIFY_SOURCE=2 -Wstrict-aliasing=2 -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Werror=date-time -Werror=format-security -nostdlibinc -msoft-float -march=armv8-a -mfloat-abi=softfp -mfpu=neon-fp-armv8 -mcpu=cortex-a53 -mfpu=neon-fp-armv8 -D__ARM_FEATURE_LPAE=1 -D__LIBC_API__=10000 -D__LIBM_API__=10000 -D__LIBDL_API__=10000 -Iexternal/libcxx/include -Iexternal/libcxxabi/include -Ibionic/libc/include -Isystem/core/include -Isystem/media/audio/include -Ihardware/libhardware/include -Ihardware/libhardware_legacy/include -Ihardware/ril/include -Iframeworks/native/include -Iframeworks/native/opengl/include -Iframeworks/av/include -isystem bionic/libc/include -isystem bionic/libc/kernel/uapi -isystem bionic/libc/kernel/uapi/asm-arm -isystem bionic/libc/kernel/android/scsi -isystem bionic/libc/kernel/android/uapi -Ilibnativehelper/include_jni -Wall -Werror -target armv7a-linux-androideabi10000 -Bprebuilts/gcc/linux-x86/arm/arm-linux-androideabi-4.9/arm-linux-androideabi/bin -fPIE -D_USING_LIBCXX -std=c++17 -Wsign-promo -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -Wno-thread-safety-negative -Wno-gnu-include-next -fvisibility-inlines-hidden -fsized-deallocation -fexceptions -fcoroutines-ts -Wno-format-zero-length -Wno-implicit-fallthrough -Wno-non-virtual-dtor -Wno-return-stack-address -Wno-unused-local-typedef -UNDEBUG -O0 -frtti -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -Werror=address-of-temporary -Werror=return-type -Wno-tautological-constant-compare -Wno-tautological-type-limit-compare -Wno-tautological-unsigned-enum-zero-compare -Wno-tautological-unsigned-zero-compare -Wno-enum-compare -Wno-enum-compare-switch -Wno-c++98-compat-extra-semi -Wno-return-std-move-in-c++11 -o /tmp/libcxx-testsuite-XQdB1M/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/Output/copy_file_large.pass.cpp.o /usr/local/google/home/enh/aosp/external/libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_large.pass.cpp -DLIBCXX_FILESYSTEM_STATIC_TEST_ROOT="/data/local/tmp/libcxx/static_test_env" -DLIBCXX_FILESYSTEM_DYNAMIC_TEST_ROOT="/data/local/tmp/libcxx/dynamic_test_env" -DLIBCXX_FILESYSTEM_DYNAMIC_TEST_HELPER="/system/bin/filesystem_dynamic_test_helper" && prebuilts/clang/host/linux-x86/clang-r353983d/bin/clang++ out/soong/.intermediates/bionic/libc/crtbegin_dynamic/android_arm_armv8-a_cortex-a73_core/crtbegin_dynamic.o /tmp/libcxx-testsuite-XQdB1M/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/Output/copy_file_large.pass.cpp.o out/soong/.intermediates/external/libcxx/libc++experimental/android_arm_armv8-a_cortex-a73_core_static/libc++experimental.a out/soong/.intermediates/external/libcxx/libc++fs/android_arm_armv8-a_cortex-a73_core_static/libc++fs.a out/soong/.intermediates/external/libunwind_llvm/libunwind_llvm/android_arm_armv8-a_cortex-a73_core_static/libunwind_llvm.a -Wl,--start-group prebuilts/clang/host/linux-x86/clang-r353983d/lib64/clang/9.0.4/lib/linux/libclang_rt.builtins-arm-android.a prebuilts/gcc/linux-x86/arm/arm-linux-androideabi-4.9/arm-linux-androideabi/lib/libatomic.a out/soong/.intermediates/build/soong/libgcc_strippe
Command: ['adb', 'shell', 'cd /data/local/tmp/libcxx/temp.nHNbnx && env LD_LIBRARY_PATH=/usr/local/google/home/enh/aosp/out/host/linux-x86/lib:/usr/local/google/home/enh/aosp/out/host/linux-x86/lib64 LIBCXX_FILESYSTEM_DYNAMIC_TEST_ROOT=/data/local/tmp/libcxx/dynamic_test_env su shell timeout 300 /data/local/tmp/libcxx/temp.nHNbnx/libcxx_test.exe; echo x$?']
Exit Code: 1
Standard Error:
--
Test case FAILED with uncaught exception:
    last checkpoint near /usr/local/google/home/enh/aosp/external/libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_large.pass.cpp::46 large_file

Summary for testsuite filesystem_copy_file_test_suite:
    0 of 1 test cases passed.
    1 of 1 assertions passed.
    0 unsupported test cases.
--

Compiled test failed unexpectedly!
********************
```

and the other failure is just a failure to clean up because of permissions.

what i _haven't_ tried yet is running this against jellybean. atm i'm using system(3) for `rm -rf` and `mkdir -p`, and i'm not sure how far back that works. (i just wanted to see "is this a realistic path we could take?", to which the answer seems to be "yes".)

**Comment by enh-google** [comment]
(since @DanAlbert never claimed his jellybean device back, a quick `adb shell` tells me that mkdir is fine but `rm` doesn't have `-f` until kitkat.)

**Comment by enh-google** [comment]
here's a C++ helper that pretends to be the python script: https://android-review.googlesource.com/c/platform/external/libcxx/+/1108472

and here's a helper-less POSIX implementation: https://android-review.googlesource.com/c/platform/external/libcxx/+/1108875

**Comment by DanAlbert** [comment]
Cool, skipping the helpers cuts out a lot of the work on the test runner side. The rest here should be fairly straightforward (assuming there are important things missing from Jelly Bean).

**Comment by alexcohn** [comment]
> If you're on Windows, you'll need a VM for building the NDK

WSL v.1 (Ubuntu) works fine with that, and much easier to set up than some VirtualBox.

**Comment by gitoby** [comment]
Thanks for all the information you provided!

We are going to look into the test setup this week. If anything is unclear, we might come back to you with questions in the next days. Otherwise, we'll estimate and give you feedback based on our initial test setup.

One thing I can already say with some certainty is that we don't have a Jelly Bean device for testing. Could you run the final tests on your side if we are able to provide a patch?

**Comment by DanAlbert** [comment]
> One thing I can already say with some certainty is that we don't have a Jelly Bean device for testing. Could you run the final tests on your side if we are able to provide a patch?

Yeah, I figured that might be the case. Shouldn't be a problem.

**Comment by enh-google** [comment]
i was hoping to work on that with danalbert this week, but food poisoning disagrees.

one vaguely related thing i did manage to do from my sickbed, not requiring any collaboration, that i should probably record _somewhere_ was that i built a static toybox and ran the toybox tests on jellybean. there were some failures (most notably anything requiring time zone information), but only about 100 failures from 1200 tests, iirc. running the tests was impossible without bootstrapping an extra bin directory on $PATH, full of symlinks to the shiny new toybox. so little actually works "right" on devices as old as jellybean... i'll borrow a kitkat device when i'm back and see what the results are like there (where i think time zone data should be compatible with the current implementation). but it does seem like this might be somewhat useful to anyone testing on old devices. the most obvious thing that's still horrifically broken is that `adb shell` doesn't return the actual exit status of the called command on ancient devices.

**Comment by Maddimax** [comment]
Hi,
I'm working with @gitoby on this issue. We've spend the last few days getting up to speed on how to work on the NDK. 

We've created two PRs:
* https://android-review.googlesource.com/c/platform/external/libcxx/+/1110835 
* https://android-review.googlesource.com/c/platform/ndk/+/1110817

The libcxx changes are based on @enh-google filesystem_test_helper.h, but moved over to the master-ndk branch. The ndk changes are pretty hacky right now, especially the upload part.

So far the test data is pushed to the device, and a couple of tests pass. Before we continue on this I'd like to ask you two to have a look if we are moving in the right direction.

Hope you are feeling better soon @enh-google!

**Comment by enh-google** [comment]
nice work, thanks! i've made a few comments on the CLs, but they're aimed more at danalbert than you.

> and a couple of tests pass

i'd ask you _why_ the failing tests fail, but having run the libc++ tests myself last week and seen how little information they provide when they fail...

but am i reading this right and you're saying you see `Expected Passes : 2` rather than 183?

**Comment by DanAlbert** [comment]
Awesome, thanks so much for looking at this!

Also left a handful of comments, but the general approach LGTM. I'll see if I can find some time today to run against JB and make sure there aren't any big issues there.

**Comment by enh-google** [comment]
i had a poke about this evening at the statvfs issue. couple of options:

1. inline headers with the statvfs reimplemented in terms of statfs, which has been there forever. (i'd actually implemented the full syscall(3) nonsense before i realized that!)

2. `#if __ANDROID__` in libc++ to use statfs instead of statvfs.

why can't we just switch libc++ over to statfs, given that's basically what #1 does? because on darwin statfs and statvfs aren't trivially convertible. in particular, darwin statfs is a subset of the stuff libc++ needs from statvfs.

i'm leaning towards #1 on the assumption that it's easier to maintain (given that upstream could add new uses of statvfs), even if i don't think -- for linux code -- there's much value to statvfs. but someone with code to port between iOS and Android might find it useful.

**Comment by DanAlbert** [comment]
Option 1 SGTM

**Comment by enh-google** [comment]
statvfs for API < 19: https://android-review.googlesource.com/c/platform/bionic/+/1111457

**Comment by Maddimax** [comment]
Thanks for all the feedback, we are currently going through it all. There is one test failing looking for "<cuchar>", from what I understand its basically an alias for "<uchar.h>". Would it be correct to create a file in bionic that just includes uchar.h ?

**Comment by Maddimax** [comment]
> nice work, thanks! i've made a few comments on the CLs, but they're aimed more at danalbert than you.
> 
> > and a couple of tests pass
> 
> i'd ask you _why_ the failing tests fail, but having run the libc++ tests myself last week and seen how little information they provide when they fail...
> 
> but am i reading this right and you're saying you see `Expected Passes : 2` rather than 183?

I just reran the full filesystems/* testsuite and got:
libc++: PASS 254/258 FAIL 4/258 SKIP 0/258

**Comment by Maddimax** [comment]
We've updated the PRs:

* Changed statvfs calls to statfs calls for ANDROID < 19 ( @enh-google changes in bionic did not seem to apply when I tried to checkbuild again as the statvfs header used came from a prebuilt? ) 
* Fixed remaining test failures in filesystems
* Reverted changes in ndk/test/types.py
* Changed to use ndk.paths.ANDROID_DIR instead of ../.. 

We've tried running the test for android-16 ( x86, instead of x86_64 ), and are running into strange errors. For instance:

```
filesystem_error::what() does not match expected output:
  expected: "filesystem error: in file_size: No such file or directory [""]"
  actual:   "ted on an object without an associated state."
```

( See the cut-off "actual" part )

a bunch of the errors seems to be string comparison related. Is it possible that there is some sort of incompatibility between the C++ implementations somewhere? That somehow while compiling the tests it sees a different version of std::string compared to the actual libcxx version?

**Comment by DanAlbert** [comment]
> Changed statvfs calls to statfs calls for ANDROID < 19 ( @enh-google changes in bionic did not seem to apply when I tried to checkbuild again as the statvfs header used came from a prebuilt? )

Yeah, bionic (and the rest of the sysroot) is a prebuilt artifact for the NDK's build. The sysroot comes from various source branches depending on what stage of development the OS is in (i.e. beta versions come from an internal branch rather than AOSP, since AOSP doesn't have the beta APIs yet).

I'll get that updated so you can drop that part of the libc++ change. Both work, but since it might be useful for portability of other code as well we might as well do it in bionic.

> Is it possible that there is some sort of incompatibility between the C++ implementations somewhere? That somehow while compiling the tests it sees a different version of std::string compared to the actual libcxx version?

Things usually end up a lot more broken when that happens, but it's hard to rule out. I'll re-test things on android-16 Arm and see if it's present there. It might be a few days before I have a chance, but I'll try to get to it as soon as I find the time.

**Comment by DanAlbert** [comment]
https://android-review.googlesource.com/c/platform/prebuilts/ndk/+/1113492 has the update you need. Can cherry-pick it to your workspace if it hasn't been submitted by then (the "Download" button at the top of the files section has the command that will cherry-pick from gerrit).

**Comment by DanAlbert** [comment]
I can upload a set of changes that you can cherry-pick tomorrow. I was hoping to just have them submitted before it became an issue, but we're waiting on some additional changes that the kernel needs before the update gets checked in.

**Comment by Maddimax** [comment]
That would be great. I'll not be able to test them before monday/tuesday though, so no rush.

**Comment by DanAlbert** [comment]
These two patches fix it:

* https://android-review.googlesource.com/c/platform/ndk/+/1122877
* https://android-review.googlesource.com/c/platform/prebuilts/ndk/+/1122719

They'll hopefully be merged by the time you need them, but otherwise you can cherry-pick those (you'll also need to sync to pick up the new prebuilts).

**Comment by rcdailey** [comment]
Is there a way I can build my own `libc++fs` and bundle it in my APK until it gets properly put into the NDK? Ideally just a simple CMake project I can pull in and do an `add_subdirectory()` to compile it with my source.

**Comment by DanAlbert** [comment]
You could follow the subset of the instructions for building in https://github.com/android/ndk/issues/609#issuecomment-524434924.

**Comment by Maddimax** [comment]
@rcdailey You can also just include the filesystem source files directly in your project ( you could use the versions from here: https://github.com/AshampooSystems/boden/tree/master/framework/foundation/platforms/android/filesystem

**Comment by Maddimax** [comment]
@DanAlbert I finally found some time to get back to this.

There are now two issues left:
1) On Android 16, ::realpath() does not fail if the path does not exists. This can be solved by just checking "exists" in std::filesystem::__canonical, but fixing realpath() is probably the better way.
2) On Android 16, std::filesystem::permissions() when called with a symlink path, changes the linked-to files permissions when it should not. It seems like ::fchmodat does not honor AT_SYMLINK_NOFOLLOW.

Since these functions are in bionic (I guess), I don't know how to fix them as I cannot build it.

**Comment by enh-google** [comment]
i don't remember the realpath() issue you mention, but we did update realpath() in 2013 and 2014 (so this is probably fixed either in JB-MR2 or K). realpath is quite a bit of code, which is presumably why libc++ just reuses the libc implementation rather than reimplementing it itself :-(

for fchmodat(), the kernel doesn't support `AT_SYMLINK_FOLLOW` directly, so it needs to be faked in userspace. we did that back in 2015 for L (https://android-review.googlesource.com/c/platform/bionic/+/127908). that's small enough (and our code rather than BSD code) so we probably could just copy & paste that in.

**Comment by lucaszanella** [comment]
I hope it lands soon because libstd filesystem is very useful!

**Comment by Consti10** [comment]
As far as I can tell std::filesystem is not experimental in c++17 anymore
But even tough the android ndk claims to support c++17,  using e.g. `std::filesystem::exists` results in the following compilation issue (Win10, latest android studio usw)

 ```
Build command failed.
  Error while executing process C:\Users\Consti10\AppData\Local\Android\Sdk\cmake\3.6.4111459\bin\cmake.exe with arguments {--build C:\Users\Consti10\Desktop\FPV_VR_for_wifibroadcast\RenderingX\RenderingXExample\.cxx\cmake\debug\arm64-v8a --target example-renderer2}
  [1/1] Linking CXX shared library ..\..\..\..\build\intermediates\cmake\debug\obj\arm64-v8a\libexample-renderer2.so
  FAILED: cmd.exe /C "cd . && C:\Users\Consti10\AppData\Local\Android\Sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe  --target=aarch64-none-linux-android21 --gcc-toolchain=C:/Users/Consti10/AppData/Local/Android/Sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64 --sysroot=C:/Users/Consti10/AppData/Local/Android/Sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -fno-addrsig -Wa,--noexecstack -Wformat -Werror=format-security  -std=c++17 -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a -static-libstdc++ -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments -Wl,-z,noexecstack -shared -Wl,-soname,libexample-renderer2.so -o ..\..\..\..\build\intermediates\cmake\debug\obj\arm64-v8a\libexample-renderer2.so CMakeFiles/example-renderer2.dir/src/main/cpp/example2/example_distortion.cpp.o  ../../../../build/intermediates/cmake/debug/obj/arm64-v8a/libGLPrograms.so -llog -landroid -llog -lEGL -lGLESv2 ../../../../build/intermediates/cmake/debug/obj/arm64-v8a/libSuperSync.so ../../../../../RenderingXCore/libs/google/gvr/jni/arm64-v8a/libgvr.so ../../../../build/intermediates/cmake/debug/obj/arm64-v8a/libTime.so -llog -landroid -llog -lEGL -lGLESv2 -latomic -lm && cd ."
  CMakeFiles/example-renderer2.dir/src/main/cpp/example2/example_distortion.cpp.o: In function `std::__ndk1::__fs::filesystem::exists(std::__ndk1::__fs::filesystem::path const&)':
  
  C:/Users/Consti10/AppData/Local/Android/Sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64/sysroot/usr/include/c++/v1/filesystem:1611: undefined reference to `std::__ndk1::__fs::filesystem::__status(std::__ndk1::__fs::filesystem::path const&, std::__ndk1::error_code*)'
```

**Comment by M2tM** [comment]
I recently encountered this when trying out a point release of NDK 20 downloaded through VS2019's Android SDK Repo. I was able to #include <filesystem> but it doesn't seem like there's full support since I get some cryptic compilation issues when trying to use any methods.

**Comment by enh-google** [comment]
yeah, we're working on it, but it won't be ready for r20 i'm afraid.

**Comment by DanAlbert** [comment]
It's much too late for r20, considering that shipped in June :)

It's also too late for r21, since that's not receiving more new features now. r22 is likely since enh has been working on this.

**Comment by M2tM** [comment]
Thanks guys! Honestly I appreciate all your hard work. I'm not sure iOS has std::filesystem support yet either so that may be a separate blocker anyway.

**Comment by oviano** [comment]
Just curious what is the latest on this as I've just come across this while trying to port an iOS application to Android. I've been using C++17 and read that Android NDK supported this, but I get these kind of link errors...

`
/Users/Oliver/Library/Android/sdk/ndk/21.0.6113669/sources/cxx-stl/llvm-libc++/include/filesystem:1620: error: undefined reference to 'std::__ndk1::__fs::filesystem::__status(std::__ndk1::__fs::filesystem::path const&, std::__ndk1::error_code*)'
`

What is the current best way to sort workaround these? I don't mind dropping some work-in-progress files into my project, but they'd conflict with what seems to be a half-implementation in NDK21?

**Comment by Apeopex** [comment]
> Just curious what is the latest on this as I've just come across this while trying to port an iOS application to Android. I've been using C++17 and read that Android NDK supported this, but I get these kind of link errors...
> 
> `/Users/Oliver/Library/Android/sdk/ndk/21.0.6113669/sources/cxx-stl/llvm-libc++/include/filesystem:1620: error: undefined reference to 'std::__ndk1::__fs::filesystem::__status(std::__ndk1::__fs::filesystem::path const&, std::__ndk1::error_code*)'`
> 
> What is the current best way to sort workaround these? I don't mind dropping some work-in-progress files into my project, but they'd conflict with what seems to be a half-implementation in NDK21?

When I started an earlier project and looking for a drop-in replacement, I have been finding good success with https://github.com/gulrak/filesystem for the past year as a drop-in replacement.

**Comment by oviano** [comment]
Thanks for that, I'll check it out!

**Comment by oviano** [comment]
Just to clarify have you been using that library on Android?

Only it states “Also there are some checks to hopefully better work on Android, but as I currently don't test with the Android NDK, I wouldn't call it a supported platform yet”

**Comment by Apeopex** [comment]
Yup, have had no issues on API 21+ for a year now.

**Comment by pkubik** [comment]
Can you confirm that using external dependency is the the only way to use `std::filesystem` today? LLVM claims that no extra libraries are required to use file system as everything is already merged in `libc++` from version 9 (https://libcxx.llvm.org/docs/UsingLibcxx.html). Also the other libs, i.e. `libc++fs` and `libc++experimental` are not available in the NDK 21.

**Comment by pkubik** [comment]
Ok, I found the roadmap that states that NDK doesn't support std::filesystem yet and it links to this issue:
```
C++ File System API

Issue 609

We don't currently build, test, or ship libc++'s std::filesystem.

There‘s a fair amount of work involved in getting these tests running. Some of the work has already been done, but there’s still more to be done for supporting older OS versions.
```
https://android.googlesource.com/platform/ndk/+/master/docs/Roadmap.md#package-management

**Comment by mateuszzz88** [comment]
In case work on tests doesn't make it until r22 - could you consider adding implementation anyway? Perhaps with opt-in / experimental flag if you are understandably not comfortable with putting not tested code in NDK?
I think that current state of having working `#include`s and missing implementation is worse. I just spent whole day migrating our project from `boost::` to `std::`. It workend on linux, it even worked on IOS, but not on android. To add insult to injury I learnt about it only after linking. Headers are not even in experimental now. 
Either way: thank you for working on this :-)

**Comment by oviano** [comment]
Yes the current implementation is irritating - either don't include it at all, or include it so it links!

As @Apeopex mentioned further up, there is a workable Android alternative which I am using and haven't had any issues with at all.

 https://github.com/gulrak/filesystem

**Comment by mateuszzz88** [comment]
@oviano If I am reading docs right, this is equivalent of using `boost::filesystem`, right? It doesn't pretend to be `std::`?

**Comment by oviano** [comment]
Yes that's correct, but I imagine the advantage is that it's a single header file so wouldn't bring in a dependency on boost which it sounds like you are moving away from.

The way the developer has implemented it means that you would use the namespace "fs" instead of std: and then you can configure it to either use his implementation OR std:filesystem.

So in my project, I use std: for all platforms except Android, but throughout the code I am referring to fs::create_directory, fs::open_file etc.

I've obviously not tried every single piece of functionality, only the basics really, but it seems like a pretty serious attempt to implement filesystem on C++11 onwards and is certainly good enough for my uses.

**Comment by DanAlbert** [comment]
> In case work on tests doesn't make it until r22 - could you consider adding implementation anyway?

That would send a strong signal that it's ready for use, and it's not. We're not going to add the library until we're reasonably certain that it works. Removing the headers until it's ready is probably an option. If std::filesystem isn't ready for r22 we'll try that.

**Comment by enh-google** [comment]
i'd rather put in a `#error` if someone's building for < 21. we do have _some_ confidence in this library because we're using it in the platform already. (and we're not running the tests there either.) to some extent you're always at the mercy of the OEM's libc and kernel changes anyway, so running the tests on a Pixel device still wouldn't be a guarantee for an NDK user. (getting the tests into CTS would help, but that isn't something we're ever likely to manage.)

i worry about letting the perfect be the enemy of the good here. (the older situation of "this isn't stabilized yet" was different, but r22 will be on a version of libc++ where it's all kosher, right?)

**Comment by DanAlbert** [comment]
> (and we're not running the tests there either.)

That isn't true. They aren't in CTS (they're not an app interface), but they're run along with the rest of the libc++ tests in the platform.

> the older situation of "this isn't stabilized yet" was different, but r22 will be on a version of libc++ where it's all kosher, right?

Until we've run the tests we don't know if the interface in the NDK is stable yet. We don't know what changes we might need to make to support, for example, large file offsets on old devices. If we find out that a break is needed there after we've shipped this as stable then we may need to rev the whole NDK libc++ ABI, which would be _extremely_ disruptive.

**Comment by DanAlbert** [comment]
Conveniently, upstream has actually fixed some of the issues that we were going to need to fix about a month ago, so there's a bit less work on our end.

After disabling a bunch of tests (everything relating to hard links, sockets, and fifos, since selinux prevents their use in /data/local/tmp), I do have a passing test run for Q devices.

There's at least one thing preventing me from running the tests on JB devices more or less at all though. I have an idea of how to fix it, but haven't had a chance to try yet. That could easily be where most of the trouble is so it's hard to say for sure whether this is close to done or not, but the few that do run have me optimistic.

**Comment by DanAlbert** [comment]
Alright, managed to get a full test run on Jelly Bean. Everything looks good within a reasonable margin of error. Tests did uncover two bugs, but they're relatively minor:

1. #1258 `std::filesystem::permissions(symlink_path, p perms, std::filesystem::perm_options::nofollow)` will not honor `nofollow` for devices older than API 23.
2. #1260 `std::filesystem::canonical` will incorrectly succeed for non-existent paths on devices older than API ~19 (not sure exactly when)

There are a fair number of other things that don't work, most of which depend on API level, but they are all related to Android security policy and `std::filesystem` signals the error appropriately:

* hardlinks aren't allowed (at least in /data/local/tmp for the shell user)
* no types of special files (`mknod`, named fifos, unix domain sockets, etc) are allowed (at least in /data/local/tmp for the shell user)
* Ability to set the permissions of a symlink is inconsistent (see #1258)

It's going to take me some time to get the test fixes pushed upstream, but if that takes too long I don't think we need to wait for that before shipping it in r22 beta 1.

As for the two bugs, I don't think those are severe enough that we necessarily need to fix them before shipping. It'd be better to get them fixed if we can, but I don't think we need to hold back the feature for them.

**Comment by enh-google** [comment]
(see https://android-review.googlesource.com/c/platform/external/libcxx/+/1110835/7/src/filesystem/operations.cpp for my libc++ workaround for those.)

**Comment by alexcohn** [comment]
Unix domain sockets are supported. So, maybe the test should run in app private sandbox, and not in `/data/local/tmp` ?

**Comment by DanAlbert** [comment]
Converting our tests to run in an app is something I've been working on. It's going to take some time, it seems, especially for the libc++ tests. Nothing untenable, just going to take longer than I want to hold this back.

Good to know that it's not wasted effort though :)

**Comment by DanAlbert** [comment]
The implementation for this is now checked in to master and will be in r22. I've forked the remaining testing cleanup into https://github.com/android/ndk/issues/1265 so I can close this to make it obvious that the part you folks care about is complete without having to read the whole thing :)

**Comment by gulrak** [comment]
>> Yes the current implementation is irritating - either don't include it at all, or include it so it links!
>>
>> As @Apeopex mentioned further up, there is a workable Android alternative which I am using and haven't had any issues with at all.
>>
>> https://github.com/gulrak/filesystem

> @oviano If I am reading docs right, this is equivalent of using `boost::filesystem`, right? It doesn't pretend to be `std::`?

These comments have been a few weeks ago, but I just found this discussion and want to add a small detail to the suggested workaround implementation in case someone might think about it:

It is comparable to `boost::filesystem` in that it is using its own namespace and so its not the same as `std::`, but it differs from `boost::filesystem` in that it pretends to be `std::filesystem` in questions of api and behavior. There are various differences between `boost::fs` and `std::fs` as for example joining absolute paths by `operator/` and the above mentioned header-only implementation tries hard to implement `std::fs` behavior, and when it differs undocumented, I handle it as a bug. The one major documented difference is, it is utf-8 oriented, so if that's your encoding, you get what `std::filesystem` would do for C++17. Details of behavior differences are in the docs.

Still, to be clear, I'm currently not developing on Android myself, and I have no CI integration for it, so the current level of support is just based on feedback and fixed issues of users. And I am thankful for that! Any feedback or help to get a CI setup on Github going to enhance support would sure be appreciated.

**Comment by hcv1027** [comment]
> The implementation for this is now checked in to master and will be in r22. I've forked the remaining testing cleanup into #1265 so I can close this to make it obvious that the part you folks care about is complete without having to read the whole thing :)

Hi @DanAlbert , can we use `std::filesystem::create_directory` in r22?

I still have the compile errors:
```bash
In function `std::__ndk1::__fs::filesystem::exists(std::__ndk1::__fs::filesystem::path const&)':
/root/workspace/3rd_party/android-ndk-r21d-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/filesystem:1615: undefined reference to `std::__ndk1::__fs::filesystem::__status(std::__ndk1::__fs::filesystem::path const&, std::__ndk1::error_code*)'
In function `std::__ndk1::__fs::filesystem::create_directory(std::__ndk1::__fs::filesystem::path const&)':
/root/workspace/3rd_party/android-ndk-r21d-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/filesystem:1555: undefined reference to `std::__ndk1::__fs::filesystem::__create_directory(std::__ndk1::__fs::filesystem::path const&, std::__ndk1::error_code*)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)

```

This is my cmake commad:
```bash
cmake -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake \
    -DANDROID_ABI=arm64-v8a \
    -DANDROID_PLATFORM=android-29 \
    -DANDROID_NATIVE_API_LEVEL=29 \
    -DANDROID_STL=c++_shared \
    -DBUILD_SHARED_LIBS=ON \
    -DCMAKE_BUILD_TYPE=RELEASE \
    -DCMAKE_INSTALL_PREFIX=$DEST \
    ..
```

Accroding to this [link](https://libcxx.llvm.org/docs/UsingLibcxx.html#using-filesystem)
I also tries to add `-lc++experimental` or `-lc++fs` in my CMakeList.txt

```cmake
target_link_libraries( vmc_io_lite
  c++fs
)
```

But this will cause below error message:
```bash
/root/workspace/3rd_party/android-ndk-r21d-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lc++fs
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

Can we use these function in r22?
`std::filesystem::create_directory`, `std::filesystem::exists`, `std::filesystem::remove_all`

**Comment by rprichard** [comment]
This path is for r21d not r22:

/root/workspace/3rd_party/android-ndk-r21d-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/filesystem

**Comment by mgood7123** [comment]
any updates on support status yet?
